apps/system/js/hardware_buttons.js

```diff
@@ -673,6 +680,63 @@
     this.hardwareButtons.setState('base', type);
   };
 
+  /**
+   * We enter the camera state when the user presses the Camera button
+   * We can fire camera or holdcamera events from this state
+   *
+   * @class HardwareButtonsCameraState
+   */
+  HardwareButtonsCameraState =
+    HardwareButtons.STATES.camera = function HardwareButtonsCameraState(hb) {
+      this.hardwareButtons = hb;
+      this.timer = undefined;
+    };
+
+  /**
+   * Entering the state.
+   * @memberof HardwareButtonsCameraState.prototype
+   */
+  HardwareButtonsCameraState.prototype.enter = function() {
+    this.timer = setTimeout(function() {
+      /**
+       * When the user holds Camera button more than HOLD_INTERVAL.
+       * @event HardwareButtonsCameraState#holdcamera
+       */
+      this.hardwareButtons.publish('holdcamera');
+      navigator.vibrate(50);
+      this.hardwareButtons.setState('base');
+    }.bind(this), this.hardwareButtons.HOLD_INTERVAL);
+  };
+
+  /**
+   * Leaving the state.
+   * @memberof HardwareButtonsCameraState.prototype
+   */
+  HardwareButtonsCameraState.prototype.exit = function() {
+    if (this.timer) {
+      clearTimeout(this.timer);
+      this.timer = undefined;
+    }
+  };
+
+  /**
+   * Process the event, maybe transition the state.
+   * @memberof HardwareButtonsCameraState.prototype
+   * @param  {String} type Name of the event to process.
+   */
+  HardwareButtonsCameraState.prototype.process = function(type) {
+    switch (type) {
+      case 'camera-button-release':
+        /**
+         * When the user releases Camera button before HOLD_INTERVAL.
+         * @event HardwareButtonsCameraState#camera
+         */
+        this.hardwareButtons.setState('base', type);
+        return;
+    }
+    console.error('Unexpected hardware key: ', type);
```

Comment:
I don't think other event types should reach here in this function... we should ignore them instead just like Line 306.